### ISSUE TITLE
Performance degradation on Xbox SX as a result of commit 

### ISSUE THREAD

**Issue by TheRhysWyrill** [libretro/RetroArch#14069]
## Description

Changes made to the D3D11/D3D12 drivers via a commit to reduce input latency has resulted in a severe impact to performance on most cores on the Xbox Series platform.

### Expected behavior

To have reduced input latency without performance being impacted.

### Actual behavior

A lot of cores are now experiencing reduced performance as shown below.

https://user-images.githubusercontent.com/27123572/174460376-868e3ed3-6e9a-4eee-a60d-96ffd662aa44.mp4

### Steps to reproduce the bug

1. Test Shadow Tower: Abyss (PS2) on an Xbox Series System.
2. Performance will drop to the 40s within the first couple minutes of the game.

### Bisect Results

This behavior appears to be a result of the changes made in this commit [31a43b9](https://github.com/libretro/RetroArch/commit/31a43b93900e9305b293b711452bfb1b5a4a177d) but as explained above it only appears to affect the Xbox for whatever reason.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: [1.10.3/[0008691](https://github.com/libretro/RetroArch/commit/0008691aa080eae4bf9acc3a7d2187e1ed265b12)]

### Environment information

- OS: Xbox Series X

**Comment by LibretroAdmin** [comment]
Hi there, I will ask @sonninnos if this can be turned into an optional setting.

**Comment by sonninnos** [comment]
Bummer. Did you try to increase the new "Max Frame Latency" option?

**Comment by TheRhysWyrill** [comment]
> Bummer. Did you try to increase the new "Max Frame Latency" option?

I have yeah, seems to still have some impact on performance with regular stutters that didn't occur beforehand but with `Max Frame Latency` on 3 it doesn't slow down to the 40s as shown above, on 1/2 is when it's especially bad.

**Comment by sonninnos** [comment]
Okeydokey, I decided to solve it by adding a fake value 0 to "Max Frame Latency", which activates this current low-latency mode, so it should cover all necessary cases without adding a separate option for it.

**Comment by TheRhysWyrill** [comment]
Awesome, thanks for providing a fix so quickly, I'll be sure to test it as soon as possible.

**Comment by LibretroAdmin** [comment]
Please test if it fixes your performance issues before closing this.

**Comment by TheRhysWyrill** [comment]
I do apologize, I made a small change to the initial commit and now it's working whether the option is disabled or enabled.

**Comment by LibretroAdmin** [comment]
Let us know if this commit as is (without any alterations made by you) fixes the performance issues. That's what we have to know.

**Comment by TheRhysWyrill** [comment]
Oh I see, in that case it does not, for some reason Xbox just really doesn't like Maximum Frame Latency set to 1 or 2, only when setting it to 3 does it mostly eliminate the performance issues outside of the odd stutter.

**Comment by sonninnos** [comment]
Oh so it wasn't completely due to the latest `WaitForVBlank` thing, but existed after adding waitable swapchains.

So shouldn't we be doing this then:
```diff
diff --git a/config.def.h b/config.def.h
index d819d3a8f4..703fe41f01 100644
--- a/config.def.h
+++ b/config.def.h
@@ -339,7 +339,11 @@
 #define DEFAULT_MAX_SWAPCHAIN_IMAGES 3

 /* D3D1x specific */
+#if defined(__WINRT__) || defined(WINAPI_FAMILY) && WINAPI_FAMILY == WINAPI_FAMILY_PHONE_APP
+#define DEFAULT_MAX_FRAME_LATENCY 3
+#else
 #define DEFAULT_MAX_FRAME_LATENCY 1
+#endif

 /* GL specific */
 #define DEFAULT_ADAPTIVE_VSYNC false
```

**Comment by TheRhysWyrill** [comment]
Seems that could be the case but yeah even with Max Frame Latency set the 3 there is the odd stutter, for example Shadow Tower Abyss is a smooth 60 with them disabled, with them enabled and set to 3 you get the odd audio pop and dips to 57/58, might just make more sense for these things to be optional via a Xbox fork as this isn't something that's effecting all users, just a small subset of users on Xbox.

**Comment by sonninnos** [comment]
Hmm, I'm not seeing any mention about the maximum in the API, so could you try editing the maximum to 4-5 and see what happens? One might think the input delay would not be worth it at some point..

**Comment by LibretroAdmin** [comment]
> Seems that could be the case but yeah even with Max Frame Latency set the 3 there is the odd stutter, for example Shadow Tower Abyss is a smooth 60 with them disabled, with them enabled and set to 3 you get the odd audio pop and dips to 57/58, might just make more sense for these things to be optional via a Xbox fork as this isn't something that's effecting all users, just a small subset of users on Xbox.

No we don't want an Xbox fork, absolutely not.

Just make any thing that degrades performance an optional setting.

**Comment by TheRhysWyrill** [comment]
> Hmm, I'm not seeing any mention about the maximum in the API, so could you try editing the maximum to 4-5 and see what happens? One might think the input delay would not be worth it at some point..
> 
> `menu/menu_setting.c` row 12549.

Setting it to 4/5 eliminates a lot of the stutters but it actually seems to perform better with it set to 4.

**Comment by sonninnos** [comment]
Ok, so if it still is not as good as without waitable swapchains altogether, then there must be a separate option, and why not also increase that maximum if it has a benefit.

**Comment by TheRhysWyrill** [comment]
> Ok, so if it still is not as good as without waitable swapchains altogether, then there must be a separate option, and why not also increase that maximum if it has a benefit.

It would seem that having waitable swapchains and "Max Frame Latency" both set to 4 has it perform identically to having the options disabled.

**Comment by TheRhysWyrill** [comment]
Fixed with commit [a0bfdcc](https://github.com/libretro/RetroArch/commit/a0bfdcce3be3eb779121d72ad6c814094959d5ed).
